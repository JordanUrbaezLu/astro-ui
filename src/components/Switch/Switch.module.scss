@use "../../utility/globals";

.switchContainer {
  align-items: center;
  background: none;
  border: 0;
  cursor: pointer;
  display: flex;
  margin: 0;
  outline: none;
  padding: 0;
}

.switchPill {
  background: gray;
  border-radius: 999px;
  height: 22px;
  position: relative;
  transition: all 0.15s ease-in;
  width: 42px;
  z-index: 1;

  &.on {
    background: globals.$secondary-border-color-hover-and-focus;
  }
}

.switchIndicator {
  background: white;
  border-radius: 999px;
  content: "";
  display: inline-block;
  height: 16px;
  position: absolute;
  right: 23px;
  top: 3px;
  transition: all 0.15s ease-in;
  width: 16px;
  z-index: 3;

  &.on {
    background-color: globals.$secondary-background-color-active;
    transform: translateX(20px);
  }
}

.switchRipple {
  background: transparent;
  border-radius: 999px;
  content: "";
  display: inline-block;
  height: 30px;
  overflow: hidden;
  position: absolute;
  right: 16px;
  top: -4px;
  transition: all 0.15s ease-in;
  width: 30px;
  z-index: 4;

  &.on {
    transform: translateX(20px);
  }
}

.switchLabel {
  color: gray !important;
  margin-left: 8px;
  transition: all 0.15s ease-in;

  &.on {
    color: globals.$secondary-text-color-hover-and-focus !important;
  }
}
