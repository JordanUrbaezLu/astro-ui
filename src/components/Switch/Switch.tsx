import classNames from "classnames";
import * as React from "react";
import styles from "./Switch.module.scss";

export type SwitchVariant = "primary" | "secondary";

export interface SwitchProps {
  /**
   * If the Switch is on
   *
   * @default false
   */
  isOn?: boolean;
  /**
   * The label for the Switch
   */
  label?: string;
  /**
   * The callback fired when the Switch is clicked
   */
  onClick: () => void;
  /**
   * The variant of the Switch
   *
   * @default "primary"
   */
  variant?: SwitchVariant;
}

export const Switch: React.FC<SwitchProps> = (props) => {
  const { isOn = false, label, onClick, variant = "primary", ...rest } = props;

  const switchPill = classNames(
    styles.switchPill,
    variant === "primary" && styles.primary,
    variant === "secondary" && styles.secondary,
    isOn && styles.on
  );

  const switchIndicator = classNames(
    styles.switchIndicator,
    variant === "primary" && styles.primary,
    variant === "secondary" && styles.secondary,
    isOn && styles.on
  );

  const switchLabel = classNames(
    styles.switchLabel,
    variant === "primary" && styles.primary,
    variant === "secondary" && styles.secondary,
    isOn && styles.on
  );

  return (
    <button
      aria-checked={isOn}
      className={switchPill}
      onClick={onClick}
      role="switch"
      type="button"
      {...rest}
    >
      <span className={switchIndicator} />
      {label && <span className={switchLabel}>{label}</span>}
    </button>
  );
};
