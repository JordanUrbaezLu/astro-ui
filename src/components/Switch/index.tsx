/* istanbul ignore file */
export * from "./Switch";
