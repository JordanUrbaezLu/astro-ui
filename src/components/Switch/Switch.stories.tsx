import { Story } from "@storybook/react";
import * as React from "react";
import { Switch, SwitchProps } from "./Switch";

export default {
  component: Switch,
  title: "Components/Switch",
};

const Template: Story<Omit<SwitchProps, "isOn" | "onClick">> = (args) => {
  const [switchState, toggleSwitchState] =
    React.useState<boolean>(true);

  return (
    <Switch
      {...args}
      isOn={switchState}
      onClick={() => toggleSwitchState(!switchState)}
    />
  );
};

export const Default = Template.bind({});

export const Primary = Template.bind({});
Primary.args = {
  variant: "primary",
};

export const Secondary = Template.bind({});
Secondary.args = {
  variant: "secondary",
};

export const PrimaryAndLabel = Template.bind({});
PrimaryAndLabel.args = {
  label: " Switch Label",
  variant: "primary",
};

export const SecondaryAndLabel = Template.bind({});
SecondaryAndLabel.args = {
  label: " Switch Label",
  variant: "secondary",
};
