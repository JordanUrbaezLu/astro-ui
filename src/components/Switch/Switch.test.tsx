import { render, screen } from "@testing-library/react";
import { Switch, SwitchVariant } from "./Switch";
import { axe } from "jest-axe";

describe("Switch", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Switch isOn onClick={jest.fn()} label=" Switch Label" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Switch isOn onClick={jest.fn()} label="Switch Label" />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render label correctly.", () => {
    const label = "Switch Label";

    render(<Switch isOn onClick={jest.fn()} label={label} />);
    expect(screen.getByRole("switch")).toHaveTextContent(label);
  });

  test.each<SwitchVariant>(["primary", "secondary"])(
    "Should render variant correctly.",
    (variant) => {
      render(
        <Switch
          isOn
          onClick={jest.fn()}
          label="Switch Label"
          variant={variant}
        />
      );
      expect(screen.getByRole("switch")).toHaveClass(variant);
    }
  );
});
