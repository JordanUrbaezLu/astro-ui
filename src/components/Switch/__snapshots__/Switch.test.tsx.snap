// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Should render component correctly. 1`] = `
<button
  aria-checked="true"
  class="switchPill primary on"
  role="switch"
  type="button"
>
  <span
    class="switchIndicator primary on"
  />
  <span
    class="switchLabel primary on"
  >
     Switch Label
  </span>
</button>
`;
